clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -778,6 +791,19 @@ constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
                                       KernelNameType)};
 }
 
+static bool hasSyclKernelHandlerArg(FunctionDecl *KernelCallerFunc) {
+  // Specialization constants in SYCL 2020 are not captured by lambda and
+  // accessed through new optional lambda argument kernel_handler
+  if (KernelCallerFunc->getNumParams() > 1)
```

Comment:
I'd rather this be a type-check as you suggest, though the attribute is in the air.  I don't feel as strongly about that anymore.